### DISCUSSION TITLE
debugging API/ problematic data

### DISCUSSION THREAD

**Discussion by rkb965** [posit-conf-2023/vetiver#33] [Q&A] [answered]
Hi, all!

I was playing with the functionality and added `debug= TRUE` and then modified the test data to have problematic data (I changed BAKERY to COFFEE SHOP).

I got this error after predicting:
```
 Error in `predict()`: 

! Failed to predict: Error in `warn_to_error()`: 
! [17, 4]: expected value in level set, but got 'COFFEE SHOP' 
Backtrace:  
1. stats::predict(endpoint, problematic_test)  
3. vetiver:::predict.vetiver_endpoint(endpoint, problematic_test)
```
 
How do I:
(1) interpret "17, 4"
(2) get results for valid data and errors for problematic data, as opposed to failing entirely when there are any errors?

Thanks!

**Comment by blairj09** [comment]
Did you submit a single observation for prediction or a collection of observations?

For question 2, are you asking if there's a way to submit a collection of observations and have the model error only on observations with problematic data and return valid predictions for observations that are problem-free? If that's the case, that isn't possible due to error handling for both R and Python.

**Reply by rkb965** [reply]
Thanks!

**Reply by rkb965** [reply]
My mind is a smidge blown by your statement about error handling. So many relevant and mysterious layers to R. Thanks for clarifying!

**Comment by juliasilge** [comment, marked as answer]
> How do I interpret "17, 4"

These are the location in the data object that are the source of the problem. You can see how it gets generated here:

``` r
many_dates <- tibble::tibble(
    x = as.Date("2021-01-01") + 0:10,
    y = as.Date("1980-02-14") + 0:10,
    z = as.POSIXct("1950-06-01", tz = "UTC") + 0:10
)

bad_data <- tibble::tibble(
    x = "potato", y = "1980-03-01", z = "1950-06-01 00:00:15"
)

readr::type_convert(bad_data, readr::as.col_spec(many_dates))
#> Warning: [0, 1]: expected date like , but got 'potato'
#> # A tibble: 1 × 3
#>   x      y          z                  
#>   <date> <date>     <dttm>             
#> 1 NA     1980-03-01 1950-06-01 00:00:15
```

<sup>Created on 2023-09-18 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

> How do I get results for valid data and errors for problematic data, as opposed to failing entirely when there are any errors?

We don't currently have a way to do this; the API errors are for the call as a whole. Like @blairj09 said, the error handling is for the whole API call.

**Reply by rkb965** [reply]
@juliasilge Thank you for spelling out these details.

Your seemingly-simple reprex has been bothering me (this looks to me like R sometimes has 0-based indexing‽‽) but I'll move my follow-up questions to Community. My only hypothesis is that the index returned from C (but why [0,1] and not [0,0]?) but I doubt that is the actual answer. But I'm sure some good folks on Community will educate me!

Thanks so much for all of your time and wisdom over conf!

``` r
char_cols <- tibble::tibble(
  x = "potato"
)
col_types = readr::as.col_spec(tibble::tibble(
  x = as.Date("2021-01-01") + 0:10)
)

guesses <- list(
  x = "character"
)

specs <- readr:::col_spec_standardise(col_types = col_types, 
                                      col_names = "x", 
                                      guessed_types = guesses)

readr:::type_convert_col(char_cols[[1]], specs$cols[[1]], which(TRUE)[1], 
                         locale_ = readr::locale(), na = c("", "NA"), trim_ws = TRUE)
#> Warning: [0, 1]: expected date like , but got 'potato'
#> [1] NA

readr:::type_convert_col
#> function (x, spec, locale_, col, na, trim_ws) 
#> {
#>     .Call(`_readr_type_convert_col`, x, spec, locale_, col, na, 
#>         trim_ws)
#> }
#> <bytecode: 0x559784b0f4b8>
#> <environment: namespace:readr>
```

<sup>Created on 2023-09-21 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

**Reply by rkb965** [reply]
For completeness, the cross-post: https://community.rstudio.com/t/does-r-sometimes-index-from-0/173913
